firedrake/supermeshing.py

```diff
@@ -0,0 +1,432 @@
+# Code for projections and other fun stuff involving supermeshes.
+import firedrake
+import ctypes
+import os
+from firedrake.supermeshimpl import assemble_mixed_mass_matrix as ammm
+from firedrake.mg.utils import get_level
+from firedrake.petsc import PETSc
+from firedrake.mg.kernels import to_reference_coordinates, compile_element
+from firedrake.utility_meshes import UnitTriangleMesh, UnitTetrahedronMesh
+from firedrake.functionspace import FunctionSpace
+from firedrake.assemble import assemble
+from firedrake.ufl_expr import TestFunction, TrialFunction
+import firedrake.mg.utils as utils
+import ufl
+from ufl import inner, dx
+import numpy
+from pyop2.datatypes import ScalarType
+from pyop2.sparsity import get_preallocation
+from pyop2.compilation import load
+from pyop2.mpi import COMM_SELF
+from pyop2.utils import get_petsc_dir
+
+
+__all__ = ["assemble_mixed_mass_matrix"]
+
+
+class BlockMatrix(object):
```

Comment:
Used to provide `matvec` for vector-/tensor-valued spaces, rather than assembling the bigger operator.